### PR TITLE
Add an extension to the Postgres resource to configure a pgAdmin 4 container.

### PR DESCRIPTION
**Pull request by andrevlins**
With this extension it is possible to add a pgAdmin container to the application.
Implemented using a lifecycle hook that adds the container if a Postgres resource exists.

### CODE REVIEW

File: src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs
Comment (davidfowl) [inline src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs:74]: Why just not just add the resource here directly?

Comment (andrevlins) [reply src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs:74]: I wanted to prevent multiple calls to WithPgAdmin from resulting in the error that a "pgAdmin" resource already exists.

I will add the container resource to WithPgAdmin itself and leave it up to the developer to ensure that it is only used once.

I'll add passing an optional resource name so the developer has control over this as well.

Comment (andrevlins) [reply src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs]: Done

File: src/Aspire.Hosting/Postgres/PgAdminContainerResource.cs
Comment (DamianEdwards) [inline src/Aspire.Hosting/Postgres/PgAdminContainerResource.cs:7]: nit: add blank line after namespace

File: src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs
Comment (DamianEdwards) [inline src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs:86]: nit:
```suggestion
        pgAdminContainer.Annotations.Add(new ContainerImageAnnotation { Image = "dpage/pgadmin4", Tag = "latest" }); 
```

File: src/Aspire.Hosting/Postgres/PgAdminContainerResource.cs
Comment (andrevlins) [reply src/Aspire.Hosting/Postgres/PgAdminContainerResource.cs:8]: done

File: src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs
Comment (andrevlins) [reply src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs]: done

File: src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs
Comment (mitchdenny) [inline src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs:91]: With PgAdmin is it possible to disable auth altogether.

File: src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs
Comment (mitchdenny) [inline src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs:88]: Is it possible to pre-configure the connection to the postgres database so devs can just click on the URL in the dashboard and everything is already connected.

Comment (davidfowl) [reply src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs:88]: That would be ideal.

**Comment by mitchdenny** [comment]
@andrevlins I am really pleased you submitted this PR. I was playing with similar concepts for MongoDB.

File: src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs
Comment (mitchdenny) [reply src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs:88]: Also add `.WithManifestPublishingCallback(ManifestPublishingCallbackAnnotation.Ignore)` as we probably don't want these containers appearing in the manifest.

Comment (davidfowl) [reply src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs:88]: ExcludeFromManifest

Comment (andrevlins) [reply src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs]: Some points to discuss about the logic of pre-configuring pgAdmin with the servers:

At the point where the implementation is done, I have no way of getting the host port associated with postgres so I had to rely on the ServiceBindingAnnotation. Consequently, if the developer does not specify a port, I have to ignore it and not generate the pre-configuration.

I believe we found the motivation to return to the previous approach of using LifecycleHook where I would have the port information already assigned and we would also gain the possibility of pre-configuring all possible Postgres servers.

Regarding how to generate the json file that needs to be mounted in the container, I am generating a temporary file for this purpose and I am not sure if this is the best option.

Finally we have the issue of having to use "host.docker.internal" as the postgres server host. I couldn't find another way and I believe it doesn't work on Linux.

File: src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs
Comment (andrevlins) [reply src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs:84]: The intention is not to have more than one pgAdmin running at the same time, for the scenario it doesn't make much sense.

If we follow the path of using LifecycleHook, the existence of this class will not be necessary.

File: src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs
Comment (andrevlins) [reply src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs]: Great, no need for any option.

File: src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs
Comment (mitchdenny) [reply src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs:106]: Looks like you've put this in now!

File: src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs
Comment (mitchdenny) [reply src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs]: Looks like this lifecycle hook is in now.

**Comment by mitchdenny** [comment]
Hello again @andrevlins ... should be getting pretty familiar with our API changes now :) This build is blocked due to the removal of the connection resource being removed.

**Comment by andrevlins** [comment]
> Hello again @andrevlins ... should be getting pretty familiar with our API changes now :) This build is blocked due to the removal of the connection resource being removed.

Fixed! :)

File: src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs
Comment (andrevlins) [reply src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs]: I did an experiment with the @davidfowl new lifecycle hook but at the point where `AfterEndpointsAllocatedAsync` is called if I add a ContainerResource to the appModel it is not executed anymore.

This would be the moment when I can get all the postgres servers to pre-configure pgadmin4 with connections to them, but I cannot initialize the pgadmin4 container at this point.

Therefore, the implementation remains as it is, and it is only possible to pre-configure pgadmin4 with the postgres server if the developer explicitly specifies a port.

What direction do you want me to take?

File: src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs
Comment (mitchdenny) [inline src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs:121]: Don't abbreviate `Temp` to `Tmp`.

**Review by mitchdenny — APPROVED** [review]
This is looking good to me. Thanks for hanging in there with this one!

**Comment by josephaw1022** [comment]
I am excited for this one. This is awesome!

**Comment by danmoseley** [comment]
@andrevlins is this one waiting on you to address the small feedback above?

File: src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs
Comment (andrevlins) [reply src/Aspire.Hosting/Postgres/PostgresBuilderExtensions.cs]: Done

**Comment by andrevlins** [comment]
> @andrevlins is this one waiting on you to address the small feedback above?

If you are talking about this comment (https://github.com/dotnet/aspire/pull/1135#discussion_r1429499618) it has already been resolved.

**Comment by danmoseley** [comment]
@mitchdenny OK to merge ?

**Comment by mitchdenny** [comment]
Thanks for the contribution @andrevlins!

**Pull request merged**
